src/main/java/seedu/address/logic/commands/AddProcCommand.java

```diff
@@ -59,13 +68,74 @@ public CommandResult execute(Model model) throws CommandException {
             throw new CommandException(Messages.MESSAGE_INVALID_CLIENT_DISPLAYED_INDEX);
         }
 
-        Client clientToAddProc = lastShownList.get(index.getZeroBased());
-        model.addProcedure(clientToAddProc, procedure);
+        Client clientToEdit = lastShownList.get(index.getZeroBased());
+        Client editedClient = null;
 
-        if (model.hasProcedureInClient(procedure)) {
+        try {
+            editedClient = addClientProcedure(clientToEdit);
+        } catch (CommandException err) {
             throw new CommandException(MESSAGE_DUPLICATE_PROCEDURE);
         }
 
+        model.setClient(clientToEdit, editedClient);
+        model.updateFilteredClientList(PREDICATE_SHOW_ALL_CLIENTS);
         return new CommandResult(String.format(MESSAGE_SUCCESS, procedure));
     }
+
+    /**
+     * Creates and returns a {@code Client} with the details of {@code clientToEdit}
+     * edited with {@code addProcedureDescriptor}.
+     */
+    private Client addClientProcedure(Client clientToEdit) throws CommandException {
+        assert clientToEdit != null;
+
+        Name updatedName = clientToEdit.getName();
+        Phone updatedPhone = clientToEdit.getPhone();
+        Email updatedEmail = clientToEdit.getEmail();
+        Plan updatedPlan = clientToEdit.getPlan();
+        Address updatedAddress = clientToEdit.getAddress();
+        Set<Tag> updatedTags = clientToEdit.getTags();
+        List<Procedure> updatedProcedures = new ArrayList<>();
+
+        try {
+            updatedProcedures.addAll(addProcedure(clientToEdit.getProcedures()));
+        } catch (CommandException err) {
+            throw new CommandException(MESSAGE_DUPLICATE_PROCEDURE);
+        }
+
+        return new Client(updatedName, updatedPhone, updatedEmail, updatedAddress, updatedPlan,
+                updatedTags, updatedProcedures);
+    }
+
+    private List<Procedure> addProcedure(List<Procedure> procedureList)
+            throws CommandException {
+
+        List<Procedure> updatedProcedureList = new ArrayList<>();
+        for (int i = 0; i < procedureList.size(); i++) {
+            if (procedureList.get(i).equals(procedure)) {
+                throw new CommandException(MESSAGE_DUPLICATE_PROCEDURE);
```

Comment:
Nice feature and clear variable naming, but maybe this part of the code can be abstracted out to another function so that it can adhere to the principle of not having arrow-head functions / deep-nesting, what do you think about this?

Reply:
Jung Bae commented in the telegram group that this will not be necessary since it consists only of a nested statement. I will be merging this pull request now!